File: frontend/ast.ts
// ast.ts - AST Node Types for Hybrid Language in TypeScript
// (c) 2022 by Joshua Lim - @joshualim30 (GitHub)
// License: MIT

// export NodeTypes
export type NodeType = 
    // Statements
    | "Program" 
    | "VariableDeclaration"
    // Expressions
    | "AssignmentExpression"
    | "NumberLiteral"
    | "Identifier" 
    | "BinaryExpression";

// export interface statements
export interface Statement { kind: NodeType; }

// export interface Expression
export interface Expression extends Statement { }

// export interface Program
export interface Program extends Statement { kind: "Program",  body: Statement[] }

// export interface VariableDeclaration 
export interface VariableDeclaration extends Statement { kind: "VariableDeclaration", constant: boolean, identifier: string, value?: Expression }

// export interface AssignmentExpression
export interface AssignmentExpression extends Expression { kind: "AssignmentExpression", assignee: Expression, value: Expression }

// export interface BinaryExpression
export interface BinaryExpression extends Expression { kind: "BinaryExpression",  left: Expression,  right: Expression,  operator: string }

// export interface Identifier
export interface Identifier extends Expression { kind: "Identifier",  symbol: string } // symbol is the name of the variable

// export interface NumberLiteral
export interface NumberLiteral extends Expression { kind: "NumberLiteral",  value: number }

File: language/src/ast.rs
// ast.rs
// Abstract Syntax Tree for the Hybrid language

/// Represents a type in the Hybrid type system
#[derive(Debug, Clone, PartialEq)]
pub enum HybridType {
    Int,
    Float,
    String,
    Bool,
    Void,
    Null,
    Array(Box<HybridType>),           // array[int]
    Map(Box<HybridType>, Box<HybridType>), // map{string, int}
}

/// A typed parameter: (name, type)
#[derive(Debug, Clone, PartialEq)]
pub struct TypedParam {
    pub name: String,
    pub param_type: HybridType,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Number(f64),
    Boolean(bool),
    StringLiteral(String),
    Identifier(String),
    Assign {
        name: String,
        value: Box<Expr>,
    },
    Array(Vec<Expr>),
    Map(Vec<(Expr, Expr)>),
    Index {
        target: Box<Expr>,
        index: Box<Expr>,
    },
    Binary {
        left: Box<Expr>,
        operator: BinaryOp,
        right: Box<Expr>,
    },
    Unary {
        operator: UnaryOp,
        operand: Box<Expr>,
    },
    FunctionCall {
        name: String,
        arguments: Vec<Expr>,
    },
    If {
        condition: Box<Expr>,
        then_branch: Box<Stmt>,
        else_branch: Option<Box<Stmt>>,
    },
    While {
        condition: Box<Expr>,
        body: Box<Stmt>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum BinaryOp {
    Add,
    Subtract,
    Multiply,
    Divide,
    Equal,
    NotEqual,
    LessThan,
    GreaterThan,
    LessThanOrEqual,
    GreaterThanOrEqual,
}

#[derive(Debug, Clone, PartialEq)]
pub enum UnaryOp {
    Negate,
    Not,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Stmt {
    Expression(Expr),
    VariableDeclaration {
        is_const: bool,
        name: String,
        var_type: HybridType,
        value: Expr,
    },
    BlockDeclaration {
        name: String,
        parameters: Vec<TypedParam>,
        return_types: Vec<HybridType>,
        body: Vec<Stmt>,
        is_foreign: bool,               // true if preceded by #lang
        foreign_lang: Option<String>,   // e.g., "python", "rust"
        raw_body: Option<String>,       // raw code for foreign blocks
    },
    Return(Option<Expr>),
    Block(Vec<Stmt>),
}

#[derive(Debug, Clone)]
pub struct Program {
    pub statements: Vec<Stmt>,
}
